python/instrumentation/openinference-instrumentation-llama-index/src/openinference/instrumentation/llama_index/_callback.py

```diff
@@ -0,0 +1,708 @@
+                # found in additional_kwargs. Thus the duplicate handling.
+                attributes.update(_get_token_counts(usage))
+        # Look for token counts in additional_kwargs of the completion payload
+        # This is needed for non-OpenAI models
+        if (additional_kwargs := getattr(response, "additional_kwargs", None)) is not None:
+            attributes.update(_get_token_counts(additional_kwargs))
+    if event_type is CBEventType.RERANKING:
+        if EventPayload.TOP_K in payload:
+            attributes[RERANKER_TOP_K] = payload[EventPayload.TOP_K]
+        if EventPayload.MODEL_NAME in payload:
+            attributes[RERANKER_MODEL_NAME] = payload[EventPayload.MODEL_NAME]
+        if EventPayload.QUERY_STR in payload:
+            attributes[RERANKER_QUERY] = payload[EventPayload.QUERY_STR]
+        if nodes := payload.get(EventPayload.NODES):
+            attributes[RERANKER_OUTPUT_DOCUMENTS if is_event_end else RERANKER_INPUT_DOCUMENTS] = [
+                {
+                    DOCUMENT_ID: node_with_score.node.node_id,
+                    DOCUMENT_SCORE: node_with_score.score,
+                    DOCUMENT_CONTENT: node_with_score.node.text,
+                    DOCUMENT_METADATA: node_with_score.node.metadata,
+                }
+                for node_with_score in nodes
+            ]
+    if EventPayload.TOOL in payload:
+        tool_metadata = cast(ToolMetadata, payload.get(EventPayload.TOOL))
+        attributes[TOOL_NAME] = tool_metadata.name
+        attributes[TOOL_DESCRIPTION] = tool_metadata.description
+        attributes[TOOL_PARAMETERS] = tool_metadata.to_openai_tool()["function"]["parameters"]
+    if EventPayload.SERIALIZED in payload:
+        serialized = payload[EventPayload.SERIALIZED]
+        if event_type is CBEventType.EMBEDDING:
+            if model_name := serialized.get("model_name"):
+                attributes[EMBEDDING_MODEL_NAME] = model_name
+        if event_type is CBEventType.LLM:
+            if model_name := serialized.get("model"):
+                attributes[LLM_MODEL_NAME] = model_name
+                invocation_parameters = _extract_invocation_parameters(serialized)
+                invocation_parameters["model"] = model_name
+                attributes[LLM_INVOCATION_PARAMETERS] = json.dumps(invocation_parameters)
+    return attributes
+
+
+def _extract_invocation_parameters(serialized: Mapping[str, Any]) -> Dict[str, Any]:
+    # FIXME: this is only based on openai. Other models have different parameters.
+    if not hasattr(serialized, "get"):
+        return {}
+    invocation_parameters: Dict[str, Any] = {}
+    additional_kwargs = serialized.get("additional_kwargs")
+    if additional_kwargs and isinstance(additional_kwargs, Mapping):
+        invocation_parameters.update(additional_kwargs)
+    for key in ("temperature", "max_tokens"):
+        if (value := serialized.get(key)) is not None:
+            invocation_parameters[key] = value
```

Comment:
Is there not a more general way of grabbing invocation parameters? I don't remember needing specific logic for OpenAI.